books/RayTracingTheRestOfYourLife.html

```diff
@@ -423,184 +571,504 @@
+number generator, where the distribution is defined by the PDF. We'll just _assume_ that there
+exists a function $f(d)$ that takes uniform input and produces a nonuniform distribution weighted by
+PDF. Now we need to figure out a way to solve for $f(d)$.
+
+For the PDF given above, where $p(r) = \frac{r}{2}$, the probability of a random sample is higher
+toward 2 than it is toward 0. More 1.8s and 1.9s should be generated than 0.1s and 0.2s. If we put
+aside our mathematics hat for a second and put on our computer science hat, maybe we can figure out
+a smart way of partitioning the PDF. We know that there is a higher probability near 2 than near
+0, but what is the value that splits the probability in half? What is the value that a number chosen
+at random has a 50% chance of being higher than and 50% chance of being lower than? What is the $x$
+that solves:
+
+    $$ 50\% = \int_{0}^{x}  \frac{r}{2} dr  = \int_{x}^{2}  \frac{r}{2} dr $$
+
+Solving gives us:
+
+    $$ 0.5 = \frac{r^2}{4} \Big|_{0}^{x} $$
+    $$ 0.5 = \frac{x^2}{4} $$
+    $$ x^2 = 2$$
+    $$ x = \sqrt{2}$$
+
+As a crude approximation we could create a function `f(d)` that takes `double d = random_double()`.
+It checks to see if `d` is less than 0.5 or greater. If it is less than 0.5 it produces a uniform
+number between 0 and $\sqrt{2}$, if it is greater than 0.5 it produces a uniform number between
+$\sqrt{2}$ and 2.
 
-  $$ P(1.0) = \frac{1}{4} $$
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    double function f(double d)
+    {
+        if(d <= 0.5)
+        {
+            return sqrt(2.0) * random_double();
+        }
+        else
+        {
+            return sqrt(2.0) + (2 - sqrt(2.0)) * random_double();
+        }
+
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [crude-approx]: A crude, first-order approximation to nonuniform PDF]
+
+While our initial random number generator was uniform from 0 to 1:
+
+   ![Figure [uniform-dist]: A uniform distribution](../images/fig-3.05-uniform-dist.jpg)
+
+Our, new, crude approximation for $\frac{r}{2}$ is nonuniform (but only just):
+
+    ![Figure [approx-f]: A nonuniform distribution for r/2](../images/fig-3.06-nonuniform-dist.jpg)
+
+We had the analytical solution to the integration above, so we could very very easily solve for the
+50% value. We could also solve for this experimentally. If we had a function that couldn't solve
+for the integration for--or didn't _want_ to solve the integration for--we could get an experimental
+result close to the real value. Let's take the function:
+
+    $$ p(x) = e^{\frac{-x}{2 \pi}} sin^2(x) $$
+
+Which looks a little something like this:
+
+    ![Figure [exp-sin2]: A function that we don't want to solve analytically](../images/fig-3.07-exp-sin2.jpg)
+
+From $0$ to $2\pi$ as example:
```

Comment:
Jumping from the text right into the full code is quite abrupt here. It would help to spend more time here to set up the problem and approach that the following code tackles, so the reader has a chance to reflect on how _they_ might solve this problem.

It takes some code reading and parsing to work out that the following code is creating an ordered list of x, f(x), and then walks the result until the accumulated sum is about halfway.

In fact, I think it would be more performant and more clear to just fill an array of {x, f(x)}, and then call `std::sort` (introduced in the BVT code) before searching for the halfway mark.